subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/AbstractSnapshotWithChildrenTest.groovy

```diff
@@ -0,0 +1,208 @@
+        newChildren.add(insertionPoint, newChild)
+        return newChildren
+    }
+
+    List<CHILD> childrenWithSelectedChildRemoved() {
+        children.findAll { it != selectedChild }
+    }
+
+    CHILD getNodeWithIndexOfSelectedChild(List<CHILD> newChildren) {
+        int index = children.indexOf(selectedChild)
+        return newChildren[index]
+    }
+
+    private int getInsertionPoint() {
+        int childIndex = SearchUtil.binarySearch(children) {
+            candidate -> PathUtil.compareFirstSegment(absolutePath, offset, candidate.pathToParent, CASE_SENSITIVE)
+        }
+        assert childIndex < 0
+        return -childIndex - 1
+    }
+
+    String getCommonPrefix() {
+        return selectedChild.pathToParent.substring(0, PathUtil.lengthOfCommonPrefix(selectedChild.pathToParent, absolutePath, offset, CASE_SENSITIVE))
+    }
+
+    String getPathFromCommonPrefix() {
+        return absolutePath.substring(offset + commonPrefix.length() + 1)
+    }
+
+    String getSelectedChildPathFromCommonPrefix() {
+        return selectedChild.pathToParent.substring(commonPrefix.length() + 1)
+    }
+
+    def getDescendantSnapshotOfSelectedChild(@Nullable MetadataSnapshot foundSnapshot) {
+        def descendantOffset = offset + selectedChild.pathToParent.length() + 1
+        1 * selectedChild.getSnapshot(absolutePath, descendantOffset, CASE_SENSITIVE) >> Optional.ofNullable(foundSnapshot)
+    }
+
+    def invalidateDescendantOfSelectedChild(@Nullable FileSystemNode invalidatedChild) {
+        def descendantOffset = offset + selectedChild.pathToParent.length() + 1
+        1 * selectedChild.invalidate(absolutePath, descendantOffset, CASE_SENSITIVE) >> Optional.ofNullable(invalidatedChild)
+    }
+
+    @SuppressWarnings("GrMethodMayBeStatic")
+    void noMoreInteractions() {
+        _ * _.pathToParent
+        0 * _
+    }
+
+    /**
+     * Different lists of relative paths of the initial children of the node under test.
+     */
+    static final INITIAL_CHILD_CONSTELLATIONS = [
+        ['name'],
+        ['name', 'name1'],
+        ['name', 'name1', 'name12'],
+        ['name', 'name1', 'name12', 'name2', 'name21'],
+        ['name/some'],
+        ['name/some/other'],
+        ['name/some', 'name2'],
+        ['name/some', 'name2/other'],
+        ['name/some', 'name2/other'],
+        ['name', 'name1/some', 'name2/other/third'],
+        ['aa/b1', 'ab/a1', 'name', 'name1/some', 'name2/other/third'],
+        ("a".."z").toList(),
+    ]
+
+    /**
+     * The queried/updated path has no common prefix with any of the initial children of the node under test.
+     *
+     * E.g.
+     *   path: 'name0/some
+     *   children: ['name/some', 'other']
+     * or
+     *   path: 'completelyDifferent'
+     *   children: ['name/some', 'other']
+     */
+    static final List<VirtualFileSystemTestSpec> NO_COMMON_PREFIX = INITIAL_CHILD_CONSTELLATIONS.collectMany { childPaths ->
```

Comment:
IIUC what's going on here is we are manipulating a path that is not a descendant of any of the already known children, right?

Reply:
Yeah, we make sure that the first path segment of the selected path is different from any of the children.